src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,80 @@
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
+    private static final String EXCEPTION_MESSAGE_COMPONENT_ANNOTATION
+            = "We can't make an instance of this class!";
+    private static final String EXCEPTION_MESSAGE_SET_FIELD
+            = "Can't initialize field value. ";
+    private static final String EXCEPTION_MESSAGE_NEW_INSTANCE
+            = "Can't create a new instance of ";
     private static final Injector injector = new Injector();
+    private final Map<Class<?>, Object> instances = new HashMap<>();
 
     public static Injector getInjector() {
         return injector;
     }
 
     public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+        if (!interfaceClazz.isAnnotationPresent(Component.class)) {
+            throw new RuntimeException(EXCEPTION_MESSAGE_COMPONENT_ANNOTATION);
+        }
+        Object clazzImplementationInstance = null;
+        Class<?> clazz = findImplementation(interfaceClazz);
+        Field[] declaredFields = clazz.getDeclaredFields();
+        for (Field field : declaredFields) {
+            if (field.isAnnotationPresent(Inject.class)) {
+                Object fieldInstance = getInstance(field.getType());
+                clazzImplementationInstance = createNewInstance(clazz);
+                field.setAccessible(true);
+                try {
+                    field.set(clazzImplementationInstance, fieldInstance);
+                } catch (IllegalAccessException e) {
+                    throw new RuntimeException(EXCEPTION_MESSAGE_SET_FIELD, e);
+                }
+            }
+        }
+        if (clazzImplementationInstance == null) {
+            clazzImplementationInstance = createNewInstance(clazz);
+        }
+        return clazzImplementationInstance;
+    }
+
+    private Object createNewInstance(Class<?> clazz) {
+        if (instances.containsKey(clazz)) {
+            return instances.get(clazz);
+        }
+        Object instance;
+        try {
+            Constructor<?> constructor = clazz.getConstructor();
+            instance = constructor.newInstance();
+        } catch (InstantiationException | IllegalAccessException
+                 | InvocationTargetException | NoSuchMethodException e) {
+            throw new RuntimeException(EXCEPTION_MESSAGE_NEW_INSTANCE + clazz.getName());
+        }
+        instances.put(clazz, instance);
+        return instance;
+    }
+
+    private Class<?> findImplementation(Class<?> interfaceClazz) {
+        Map<Class<?>, Class<?>> interfaceImplementations = new HashMap<>();
+        interfaceImplementations.put(FileReaderService.class, FileReaderServiceImpl.class);
+        interfaceImplementations.put(ProductParser.class, ProductParserImpl.class);
+        interfaceImplementations.put(ProductService.class, ProductServiceImpl.class);
+        if (interfaceClazz.isInterface()) {
+            return interfaceImplementations.get(interfaceClazz);
+        }
```

Comment:
I'm fully agreed about flow. That we have to fill out the map and the `findImplementation()` method separately. And with your comms, that Injector should be abstracted from this information. But I found in common mistakes that for this task it will be ok just initialize our Map with method `Map.of()`.

https://github.com/mate-academy/jv-dependency-injection/blob/main/checklist.md#:~:text=Make%20Interface%20Implementations%20map%20a%20class%20field.%20You%20can%20fill%20it%20in%20using%20Map.of()